lib/Doctrine/MongoDB/Query/Query.php

```diff
@@ -174,20 +174,34 @@ public function execute()
                 $queryOptions = $this->getQueryOptions('new', 'select', 'sort', 'upsert');
                 $queryOptions = $this->renameQueryOptions($queryOptions, ['select' => 'fields']);
 
-                return $this->collection->findAndUpdate(
-                    $this->query['query'],
-                    $this->query['newObj'],
-                    array_merge($options, $queryOptions)
-                );
+                $collection = $this->collection;
+                $query = $this->query;
+
+                $closure = function() use ($collection, $query, $options, $queryOptions) {
```

Comment:
FWIW, you can use `$this` in closures (support for binding `$this` automatically was added in PHP 5.4).